test/unit/fact_cleaner_test.rb

```diff
@@ -0,0 +1,38 @@
+require 'test_helper'
+
+class FactCleanerTest < ActiveSupport::TestCase
+  let(:cleaner) do
+    cleaner = FactCleaner.new
+    cleaner.instance_variable_set '@to_delete', []
+    cleaner
+  end
+
+  def setup
+    @root_1_fact = FactoryGirl.create(:fact_name)
+    @root_2_fact = FactoryGirl.create(:fact_name, :compose => true)
+
+    @root_3_fact = FactoryGirl.create(:fact_name, :compose => true)
+    @root_3_child_fact = FactoryGirl.create(:fact_name, :parent_id => @root_2_fact.id)
+
+    @root_4_fact = FactoryGirl.create(:fact_name, :compose => true)
+    @root_4_child_1_fact = FactoryGirl.create(:fact_name, :parent_id => @root_4_fact.id)
+    @root_4_fact_value = FactoryGirl.create(:fact_value, :fact_name => @root_4_child_1_fact)
+    @root_4_child_2_fact = FactoryGirl.create(:fact_name, :parent_id => @root_4_fact.id)
+
+    @root_5_fact = FactoryGirl.create(:fact_name, :compose => true)
+
+    @root_6_fact = FactoryGirl.create(:fact_name)
+    @root_6_fact_value = FactoryGirl.create(:fact_value, :fact_name => @root_6_fact)
+  end
+
+  test "searching orphaned root facts" do
+    cleaner.clean!
+    remaining = FactName.where(:name => [@root_1_fact, @root_2_fact, @root_3_fact, @root_4_fact,
+                                         @root_5_fact, @root_3_child_fact, @root_4_child_1_fact,
+                                         @root_4_child_2_fact, @root_6_fact].map(&:name))
+    assert_include remaining, @root_4_fact
```

Comment:
Before I split it into contexts specifying befores etc I converted it into separate tests which makes it more readable. Coverage remains the same. As noted earlier, most of tests in Foreman do not use spec syntax capabilities, I made the mistake once and then had to rewrite all role tests from spec to testunit syntax :-)